feat: 공통컴포넌트 select 개발 (#23)

yeeun-kor opened the pull request:
<!-- PR 제목 규칙:
[type]: 작업 요약 (#이슈번호)
예시: feat: 로그인 페이지 UI 구현 (#12)
-->

## #️⃣ 연관된 이슈
- closes #23

## 📑 구현 사항
- props타입 선언 
- Select컴포넌트 1개만 선언하고 필요한 옵션부분을 props를 전달하도록 구조 변경하였음 
- 사요 방법은 아래 참조 

## 🌀 어려웠던 점 / 고민했던 부분
- 처음에는 Select컴포넌트를 어떻게 구현하면 재사용성을 높여줄 수 있을까 고민이 많았습니다. 
- 원래는 Select컴포넌트 공통 컴포넌트로 구현화 시키는게 감이 안잡혔으나, 지훈님의 조언으로 props로 받아서 공통으로 묶어버렸습니다. 


## 📸 구현 이미지 (선택)
### 구현예제
![select](https://github.com/user-attachments/assets/849791af-590c-453f-bd55-0f365e5081e0)



### 아이템 클릭시 value값으로 콘솔창 출력 
<img width="2046" height="942" alt="image" src="https://github.com/user-attachments/assets/a9580d0f-1eb2-4551-bff5-7253ffbeaee3" />
- api문서 41번째줄 참고하였습니다.

## ❇️ 코드 설명 (선택)
### 타입 설명
```tsx
type SelectProps = {
  data: SelectData
  placeHolder: string
  title?: string
  icon?: React.ReactNode //아이콘 추가함
  onValueChange?: (value: string) => void
}
```
- 아이템 항목으로 뿌려질 친구들을 itemValue,itemText의 배열로 받아올 수 있게 설정하였습니다. 
- placeHolder 로 개발자가 직접 타이핑 가능합니다
- title은 드롭다운 윗 부분 텍스트 부분 입니다 
- onValueChange 는 사용자가 선택한 값을 부모쪽으로 전달해주는 함수 입니다 
- 선택적으로 아이콘을 추가할 수 있습니다.

### 사용예시
```tsx
      {/* 아이콘있고 타이틀 없는 버전 */}
      <Select
        icon={<Folder />}
        data={categoryData}
        placeHolder="전체카테고리"
        onValueChange={(e) => console.log(e)} //디버깅
      ></Select>

      {/* 아이콘없고 타이틀 있는 버전 */}
      <Select title="정렬" placeHolder="최신순" data={sortData} />
```
- 몫데이터 예시에서 서버의 응답데이터를 배열로 저장했습니다 
- 선택적으로 아이콘이 있고, 타이틀이 없는 버전 사용 가능하고 
- 아이콘이 없고, 타이틀이 있는 버전으로도 사용이 가능하도록 재사용성 높게 옵셔널로 타입선언을 하였습니다. 

## 💬 리뷰 요청사항 (선택)
> 리뷰어에게 피드백을 받고 싶은 지점이 있다면 작성해주세요.
1. 서버가 저희한테 응답해주는 드롭다운 응답값이 뭔지 궁금하네요 , api문서 어디를 참고하면 좋을까요 
2. select 태그 사용예시 확인해주시고 여기서 더 필요로하는 부분이 있다면 말씀해주세요 !

yeeun-kor committed:
chore: shadcn select 패키지 설치 및 스타일 수정
- 최신버전으로 select 패키지 설치
- border값은 회색 변수값으로 설정

yeeun-kor committed:
chore: App.css추가
- 충돌되어서 삭제된 파일 복구

yeeun-kor committed:
feat: select 공통 컴포넌트 props 제작
- props로 title과 placeholder 그리고 순회돌 데이터값도 임의로 정의하였음
- 어떻게 데이터 내려오는지, api 확인 필요
- onValueChange는 클릭시 선택되는 value값을 콘솔창에서 확인 할 수 있도록 설정함

yeeun-kor committed:
fix: 선언되지 않은 파일 삭제
- 사용하지 않는 함수 삭제 SelectLabel

agrade1 reviewed (commented):
api 에서 내려주는 옵션값 예시는 api 명세서 84번 한번 확인해보시면 좋을 것 같습니다.
공고에서 사용하는 태그 조회 API 인대 공고 목록에서 태그 기반 필터링 부분에 select가 들어가니 참고가 될거 같습니다.

질문 사항으로는 공고에서 사용하는 select의 스타일과 강의에서 사용하는 스타일이 다른대
ex) 아이콘 삽입 여부, 미 선택시 백그라운드 여부, 디자인에는 없지만 활성화된 select, option의 스타일 등에 대한 부분도 적용이 된걸까요?

yeeun-kor commented:
> api 에서 내려주는 옵션값 예시는 api 명세서 84번 한번 확인해보시면 좋을 것 같습니다. 공고에서 사용하는 태그 조회 API 인대 공고 목록에서 태그 기반 필터링 부분에 select가 들어가니 참고가 될거 같습니다.
- 넵 참고하였는데 서버에서 받아온 id값을 `itemValue` 으로 설정하면 될 것 같네요 

> 질문 사항으로는 공고에서 사용하는 select의 스타일과 강의에서 사용하는 스타일이 다른대 ex) 아이콘 삽입 여부, 미 선택시 백그라운드 여부, 디자인에는 없지만 활성화된 select, option의 스타일 등에 대한 부분도 적용이 된걸까요?
- 앗, 이 부분은 체크 못했습니다, 체크해주셔서 감사합니다 
- **아이콘을** 옵셔널로 하여 다시 수정하겠습니다. 
- **미선택시 백그라운드 여부**는 어떤걸 말씀하시는 걸까용 ? 
- **활성화된** 스타일은, hover시 백그라운드 색상 말씀하시는걸까요?
<img width="370" height="357" alt="스크린샷 2025-11-27 오후 3 20 39" src="https://github.com/user-attachments/assets/7c8809bd-c9dd-452b-b357-1d6f28689a1a" />

yeeun-kor committed:
chore: shadcn 추가 라이브러리 설치
- input,label라이브러리 설치함

yeeun-kor committed:
fix: icon추가 select
- props로 icon을 받을 수 있게 설정하였습니다.
- 선택적으로 아이콘 및 타이틀을 설정할 수 있습니다
- 몫데이터 추가하여 시각화 하였습니다

yeeun-kor commented:
@agrade1 내용 수정하여서 pr작성했습니다

yeeun-kor commented on src/mocks/selectMockData.ts:
```diff
@@ -0,0 +1,20 @@
+  { itemValue: 'high_rating', itemText: '높은 평점순' },
+  { itemValue: 'low_rating', itemText: '낮은 평점순' },
+]
+export const categoryData: SelectData = [
+  { itemValue: 'artificial-intelligence', itemText: 'AI/인공지능' },
+  { itemValue: 'Applied-ai', itemText: '응용 AI' },
+  { itemValue: 'it-programming', itemText: 'IT/프로그래밍' },
+  { itemValue: 'game-dev-all', itemText: '게임 개발' },
+  { itemValue: 'data-science', itemText: '데이터 사이언스' },
+  { itemValue: 'it', itemText: 'IT' },
+  { itemValue: 'hardware', itemText: '하드웨어' },
+  { itemValue: 'design', itemText: '디자인' },
+]
```
현재 이 부분 카테고리값을 어떻게 처리하고 보내줄지 업무 요청 보내둔 상황(백앤드5팀 )

yeeun-kor replied:
itemValue의 값은 크롤링의 결과에 따라 값이 변동될 수 있다고하니, 
변동되면 알려주신다고 함

yeeun-kor commented on src/mocks/selectMockData.ts:
```diff
@@ -0,0 +1,20 @@
+import type { SelectData } from '@/components/common/Select'
+
+export const sortData: SelectData = [
+  { itemValue: 'latest', itemText: '최신순' },
+  { itemValue: 'oldest', itemText: '오래된순' },
+  { itemValue: 'low_price', itemText: '낮은 가격순' },
+  { itemValue: 'high_price', itemText: '높은 가격순' },
+  { itemValue: 'high_rating', itemText: '높은 평점순' },
+  { itemValue: 'low_rating', itemText: '낮은 평점순' },
```
MSW몫데이터 미리 세팅해둠 . 
api명세서 41번째줄 참고하였습니다
```
    "sort": {
        type: enum(
            "latest",
            "oldest",
            "low_price",
            "high_price",
            "high_rating",
            "low_rating",
        ),
        required: False,
    },
    "category": {
        type: enum(
            "artificial-intelligence",
            "Applied-ai",
            "it-programming",
            "game-dev-all",
            "data-science",
            "it",
            "hardware",
            "design",
        ),
        required: True
    }
}
```

soohyun-bae commented on src/components/common/Select.tsx:
```diff
@@ -0,0 +1,189 @@
+    <SelectPrimitive.Portal>
+      <SelectPrimitive.Content
+        data-slot="select-content"
+        className={cn(
+          'bg-popover text-popover-foreground data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 relative z-50 ma ...
+          position === 'popper' &&
+            'data-[side=bottom]:translate-y-1 data-[side=left]:-translate-x-1 data-[side=right]:translate-x-1 data-[side=top]:-translate-y-1',
+          className
+        )}
```
이렇게 한 번에 많은 내용의 스타일들을 넣어두면 가독성측면에서 사알짝 불편함이 있는 것 같아용! cva사용해서 나눠서 관리하는 것도 좋을 것 같습니다!
[참고 자료](https://dev-102.tistory.com/entry/Tailwind-CSS-%EC%9E%98-%ED%99%9C%EC%9A%A9%ED%95%98%EA%B8%B0-with-Clxs-CVA-twMerge)

yeeun-kor replied:
헉 좋은 자료 공유와 접근법 제안 감사합니다! 오늘내로 새로운 이슈 파서 작업 해보겠습니다 감사합니다!

yeeun-kor replied:
태일윈드의 괴랄한 클래스명을 최대한 깔끔하게 정리할 수 있겠네요 힣ㅎㅎㅎㅎ

soohyun-bae commented on src/components/common/Select.tsx:
```diff
@@ -0,0 +1,189 @@
+}: SelectProps) {
+  return (
+    <div>
+      {' '}
```
이 코드는 간격을 위한 코드인걸까요??
간격을 위한거라면 스타일로 주는건 어떨까요??!!

yeeun-kor replied:
앗 이부분은 자동완성으로 들어간 것 같네요! 바로 삭제하겠습니다!

soohyun-bae reviewed (approved):
수고 많으셨습니다!!
